pyanaconda/payload/rpmostreepayload.py

```diff
@@ -186,110 +112,61 @@ def install(self):
-        # See <https://github.com/rhinstaller/anaconda/issues/1117>
-        # The first path here is used by <https://pagure.io/fedora-lorax-templates>
-        # and the second by <https://github.com/projectatomic/rpm-ostree-toolbox/>
-        if OSTree.check_version(2017, 8):
-            for path in ['/ostree/repo', '/install/ostree/repo']:
-                if os.path.isdir(path + '/objects'):
-                    pull_opts['localcache-repos'] = Variant('as', [path])
-                    break
+        from pyanaconda.modules.payloads.payload.rpm_ostree.installation import \
+            ChangeOSTreeRemoteTask
+        task = ChangeOSTreeRemoteTask(
+            data,
+            use_sysroot=False,
+            physroot=conf.target.physical_root
+        )
+        task.run()
 
         try:
-            repo.pull_with_options(data.remote,
-                                   Variant('a{sv}', pull_opts),
-                                   progress, cancellable)
-        except GError as e:
-            exn = PayloadInstallError("Failed to pull from repository: %s" % e)
+            from pyanaconda.modules.payloads.payload.rpm_ostree.installation import \
+                PullRemoteAndDeleteTask
+            task = PullRemoteAndDeleteTask(data)
+            task.progress_changed_signal.connect(self._pull_progress_cb)
+            task.run()
+
+        except PayloadInstallError as exn:
             log.error(str(exn))
             if errors.errorHandler.cb(exn) == errors.ERROR_RAISE:
                 progressQ.send_quit(1)
                 util.ipmi_abort(scripts=self.data.scripts)
                 sys.exit(1)
 
-        log.info("ostree pull: %s", progress.get_status() or "")
-        progressQ.send_message(_("Preparing deployment of %s") % (ref, ))
-
-        # Now that we have the data pulled, delete the remote for now.
-        # This will allow a remote configuration defined in the tree
-        # (if any) to override what's in the kickstart.  Otherwise,
-        # we'll re-add it in post.  Ideally, ostree would support a
-        # pull without adding a remote, but that would get quite
-        # complex.
-        repo.remote_delete(data.remote, None)
-
-        self._safe_exec_with_redirect("ostree",
-                                      ["admin", "--sysroot=" + conf.target.physical_root,
-                                       "os-init", data.osname])
-
-        admin_deploy_args = ["admin", "--sysroot=" + conf.target.physical_root,
-                             "deploy", "--os=" + data.osname]
-
-        admin_deploy_args.append(data.remote + ':' + ref)
-
-        log.info("ostree admin deploy starting")
+        gi.require_version("RpmOstree", "1.0")
+        from gi.repository import RpmOstree
+        # Variable substitute the ref: https://pagure.io/atomic-wg/issue/299
+        ref = RpmOstree.varsubst_basearch(data.ref)
         progressQ.send_message(_("Deployment starting: %s") % (ref, ))
```

Comment:
Please, move this into the task and hook the progress queue to the progress report signal of the task.